src/fu-engine.c

```diff
@@ -1223,15 +1223,26 @@ fu_engine_install_tasks (FuEngine *self,
 	/* refresh all devices in case they replugged */
 	devices_after = g_ptr_array_new_with_free_func ((GDestroyNotify) g_object_unref);
 	for (guint i = 0; i < devices->len; i++) {
-		FuDevice *old = g_ptr_array_index (devices, i);
-		FuDevice *new;
+		FuDevice *old_device = g_ptr_array_index (devices, i);
+		FuDevice *old_parent;
+		FuDevice *new_parent;
+		FuDevice *new_device;
 		g_autoptr(GError) error_local = NULL;
-		new = fu_device_list_get_by_id (self->device_list, fu_device_get_id (old), &error_local);
-		if (new == NULL) {
+		new_device = fu_device_list_get_by_id (self->device_list, fu_device_get_id (old_device), &error_local);
+		if (new_device == NULL) {
 			g_warning ("failed to find new device: %s", error_local->message);
 			continue;
 		}
-		g_ptr_array_add (devices_after, g_object_ref (new));
+		old_parent = fu_device_get_parent (old_device);
+		if (old_parent != NULL) {
+			new_parent = fu_device_list_get_by_id (self->device_list, fu_device_get_id (old_parent), &error_local);
+			if (new_parent == NULL) {
+				g_warning ("failed to find new parent: %s", error_local->message);
+				continue;
+			}
+			fu_device_set_parent (new_device, new_parent);
```

Comment:
Not sure about this; this seems the kind of thing the FuDeviceList should do. i.e. in `fu_device_list_replace()` we'd set the parent on `item->device` if it existed for `item->device_old`

Reply:
Part of the problem I'm hitting is that the parents are wrong for devices that happened earlier in the transaction (ie they're not in replug so `fu_device_list_replace` doesn't run).

I've experimented with lots of things to solve this but I'm thinking now it might just be better to not try to rely on parents as heavily in `composite_cleanup`.